rasa_core/featurizers.py

```diff
@@ -163,18 +136,425 @@ def encode(self, active_features, input_feature_map):
+
+        if not self.share_vocab:
+            self.bot_vocab = self._create_label_token_dict(bot_labels,
+                                                           self.split_symbol)
+            self.user_vocab = self._create_label_token_dict(user_labels,
+                                                            self.split_symbol)
+        else:
+            self.bot_vocab = self._create_label_token_dict(bot_labels + user_labels)
+            self.user_vocab = self.bot_vocab
+
+        self.additional_features = other_labels
+
+    def encode(self, active_features, domain):
+
+        user_labels, bot_labels, other_labels = self.labels
+
+        num_features = (len(self.user_vocab) +
+                        len(self.bot_vocab) +
+                        len(self.additional_features))
+
+        if active_features is None or None in active_features:
+            return np.ones(num_features, dtype=int) * -1
+
+        used_features = np.zeros(num_features, dtype=int)
+        for feature_name, prob in active_features.items():
+
+            if feature_name in user_labels:
+                if 'prev_action_listen' not in active_features:
+                    # we predict next action from bot action
+                    # feature_name = 'intent_listen'
+                    used_features[:len(self.user_vocab)] = 0
+                else:
+                    for t in feature_name.split(self.split_symbol):
+                        used_features[self.user_vocab[t]] += 1
+
+            elif feature_name[len('prev_'):] in bot_labels:
+                for t in feature_name[len('prev_'):].split(self.split_symbol):
+                    used_features[len(self.user_vocab) +
+                                  self.bot_vocab[t]] += 1
+
+            elif feature_name in self.additional_features:
+                if prob > 0:
+                    idx = (len(self.user_vocab) +
+                           len(self.bot_vocab) +
+                           self.additional_features.index(feature_name))
+                    used_features[idx] += 1
+            else:
+                logger.warning(
+                    "Feature '{}' could not be found in "
+                    "feature map.".format(feature_name))
+
+        return used_features
+
+
+class Featurizer(object):
+    """Base class for actual tracker featurizers"""
+    def __init__(self, featurize_mechanism=None):
+        # type: (Optional[FeaturizeMechanism]) -> None
+        if featurize_mechanism is not None:
+            self.featurize_mechanism = featurize_mechanism
+        else:
+            self.featurize_mechanism = FeaturizeMechanism()
+
+    def _pad_states(self, states):
+        return states
+
+    def _featurize_states(self, trackers_as_states, domain):
+        """Create X"""
+        features = []
+        true_lengths = []
+
+        for tracker_states in trackers_as_states:
+            if len(trackers_as_states) > 1:
```

Comment:
why do we only pad in this case?

Reply:
because during prediction `len(trackers_as_states) == 1` and since we use `dynamic_rnn` in keras, we do not need to pad